beacon-chain/sync/initial-sync/round_robin.go

```diff
@@ -0,0 +1,258 @@
+		//   i.e. the first peer is asked for block 64, 68, 72... while the second peer is asked for
+		//   65, 69, 73... and so on for other peers.
+		var request func(start uint64, step uint64, count uint64, peers []peer.ID, remainder int) ([]*eth.BeaconBlock, error)
+		request = func(start uint64, step uint64, count uint64, peers []peer.ID, remainder int) ([]*eth.BeaconBlock, error) {
+			if len(peers) == 0 {
+				return nil, errors.WithStack(errors.New("no peers left to request blocks"))
+			}
+
+			for i, pid := range peers {
+				start := start + uint64(i)*step
+				step := step * uint64(len(peers))
+				count := mathutil.Min(count, (helpers.StartSlot(finalizedEpoch+1)-start)/step)
+				// If the count was divided by an odd number of peers, there will be some blocks
+				// missing from the first requests so we accommodate that scenario.
+				if i < remainder {
+					count++
+				}
+				// asking for no blocks may cause the client to hang. This should never happen and
+				// the peer may return an error anyway, but we'll ask for at least one block.
+				if count == 0 {
+					count = 1
+				}
+				req := &p2ppb.BeaconBlocksByRangeRequest{
+					HeadBlockRoot: root,
+					StartSlot:     start,
+					Count:         count,
+					Step:          step,
+				}
+
+				resp, err := s.requestBlocks(ctx, req, pid)
+				log.WithField("peer", pid.Pretty()).Debugf("Received %d blocks", len(resp))
+				if err != nil {
+					// fail over to other peers by splitting this requests evenly across them.
+					ps := append(peers[:i], peers[i+1:]...)
+					log.WithError(err).WithField("remaining peers", len(ps)).WithField("peer", pid.Pretty()).Debug("Request failed, trying to round robin with other peers")
+					if len(ps) == 0 {
+						return nil, errors.WithStack(errors.New("no peers left to request blocks"))
+					}
+					_, err = request(start, step, count/uint64(len(ps)) /*count*/, ps, int(count)%len(ps) /*remainder*/)
+					if err != nil {
+						return nil, err
+					}
+				}
+				blocks = append(blocks, resp...)
+			}
+
+			return blocks, nil
+		}
+
+		blocks, err := request(
+			s.chain.HeadSlot()+1, // start
+			1,                    // step
+			blockBatchSize,       // count
+			peers,                // peers
+			0,                    // remainder
+		)
+		if err != nil {
+			return err
+		}
+
+		// Since the block responses were appended to the list, we must sort them in order to
+		// process sequentially. This method doesn't make much wall time compared to block
+		// processing.
+		sort.Slice(blocks, func(i, j int) bool {
+			return blocks[i].Slot < blocks[j].Slot
+		})
+
+		for _, blk := range blocks {
+			counter.Incr(1)
+			rate := float64(counter.Rate()) / counterSeconds
+			if rate == 0 {
+				rate = 1
+			}
+			timeRemaining := float64(slotsSinceGenesis(genesis)-blk.Slot) / rate
+			log.WithField(
+				"peers",
+				fmt.Sprintf("%d/%d", len(peers), len(peerstatus.Keys())),
+			).WithField(
+				"blocks per second",
+				fmt.Sprintf("%.1f", rate),
```

Comment:
Can this be formatted a bit nicer? Perhaps as a time type so that it can display minutes, seconds, etc. It's more intuitive to say 2 minutes instead of 120+ seconds.

Reply:
Awesome suggestion, thanks!